src/Juvix/Core/FromFrontend.hs

```diff
@@ -232,56 +276,125 @@ transformPatVar p = pure p
 
 -- | N.B. doesn't deal with pattern variables since HR doesn't have them.
 -- 'transformTermIR' does that.
-transformTermHR :: FE.Expression -> Env primTy primVal (HR.Term primTy primVal)
-transformTermHR (FE.Constant k) = HR.Prim <$> paramConstant k
-transformTermHR (FE.Let l) = transformSimpleLet l
-transformTermHR e@(FE.LetType _) = throwFF $ ExprUnimplemented e
-transformTermHR e@(FE.Match _) = throwFF $ ExprUnimplemented e
-transformTermHR (FE.Name n) = pure $ HR.Elim $ HR.Var n
-transformTermHR (FE.Lambda l) = transformSimpleLambda l
-transformTermHR (FE.Application (FE.App f xs)) = do
-  f' <- toElim f =<< transformTermHR f
-  HR.Elim . foldl HR.App f' <$> traverse transformTermHR xs
-transformTermHR (FE.Primitive (FE.Prim p)) = do
+transformTermHR ::
+  NameSymbol.Mod ->
+  FE.Expression ->
+  Env primTy primVal (HR.Term primTy primVal)
```

Comment:
As previous all these functions shouldn't fix the env, instead use an alias which spans the map, so I can add my own custom environment down later!